### PR TITLE
[quant] fill_ path for quantized tensors

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43304 [quant] quantized path for ConstantPadNd
* **#43303 [quant] fill_ path for quantized tensors**

Differential Revision: [D23231947](https://our.internmc.facebook.com/intern/diff/D23231947)

### CODE REVIEW

**Review by ailzhang — COMMENTED** [review]
ehhh what's the main goal of this PR (perf?) could you please add a bit context?

**Comment by z-a-f** [comment]
> ehhh what's the main goal of this PR (perf?) could you please add a bit context?

It's introducing the `fill_` for quantized tensors, which is needed for different ops (specifically for the ConstantPad)

File: test/quantization/test_quantized_tensor.py [hunk 1]
```diff
@@ -484,6 +484,32 @@ def test_qtensor_clone(self):
...
+                     [torch.int8, torch.uint8, torch.int32])
+
+        fill_with = -1
+        int_repr = int(round(fill_with / scale + zero_point))
+        for device in get_supported_device_types():
+            for qtype, dtype in dtypes:
+                q = torch._empty_affine_quantized([numel], scale=scale,
+                                                  zero_point=zero_point,
+                                                  device=device, dtype=qtype)
+                q_filled = q.fill_(-1)
```
Comment (raghuramank100) [inline test/quantization/test_quantized_tensor.py:502]: NIT: Should this be fill_with?

File: test/quantization/test_quantized_tensor.py [hunk 2]
```diff
@@ -484,6 +484,32 @@ def test_qtensor_clone(self):
...
+    def test_qtensor_fill(self):
+        numel = 10
+        scale = 0.5
+        zero_point = 10
+
+        dtypes = zip([torch.qint8, torch.quint8, torch.qint32],
+                     [torch.int8, torch.uint8, torch.int32])
+
+        fill_with = -1
+        int_repr = int(round(fill_with / scale + zero_point))
```
Comment (raghuramank100) [inline test/quantization/test_quantized_tensor.py:496]: What happens when this goes out of range for each of the dtypes?

File: test/quantization/test_quantized_tensor.py (hunk 1 above)
Comment (z-a-f) [reply test/quantization/test_quantized_tensor.py:502]: My bad -- will change before landing!

File: aten/src/ATen/native/Fill.cpp [hunk 3]
```diff
@@ -19,6 +19,11 @@ namespace {
 } // namspace
 
 Tensor& fill_out(Tensor& self, Scalar value) {
+  if (self.is_quantized()) {
+    at::Tensor out = at::ones(self.sizes()).to(kFloat) * value;
+    self.copy_(out);
```
Comment (raghuramank100) [inline aten/src/ATen/native/Fill.cpp:24]: How are out of range fill values handled?

Comment (z-a-f) [reply aten/src/ATen/native/Fill.cpp:24]: The `copy_` method already handles the logic of quantizing and taking care of the boundaries

Comment (z-a-f) [reply aten/src/ATen/native/Fill.cpp:24]: I will put an inline comment about that in here

**Review by raghuramank100 — APPROVED** [review]

File: aten/src/ATen/native/Fill.cpp [hunk 4]
```diff
@@ -19,6 +19,13 @@ namespace {
 } // namspace
 
 Tensor& fill_out(Tensor& self, Scalar value) {
+  if (self.is_quantized()) {
+    at::Tensor out = at::ones(self.sizes()).to(kFloat) * value;
+    out = out.to(self.device());
```
Comment (vkuzo) [inline aten/src/ATen/native/Fill.cpp:24]: is this needed?  Can we assert that `out.device() == self.device()` instead?

Comment (z-a-f) [reply aten/src/ATen/native/Fill.cpp:24]: We cannot assert it as the out created in this block. If we don't assign the device, we will have to trust the copy_ to move the tensors across the devices, which is an antipattern :/

**Pull request closed without merging**